cmd/oras/internal/display/track/target.go

```diff
@@ -0,0 +1,121 @@
+/*
+Copyright The ORAS Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package track
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"reflect"
+
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras-go/v2"
+	"oras.land/oras-go/v2/content"
+	"oras.land/oras-go/v2/registry"
+	"oras.land/oras/cmd/oras/internal/display"
+	"oras.land/oras/cmd/oras/internal/display/progress"
+)
+
+// Target tracks the progress of pushing ot underlying oras.Target.
+type Target struct {
+	oras.Target
+	manager      progress.Manager
+	actionPrompt string
+	donePrompt   string
+}
+
+// NewTarget creates a new tracked Target.
+func NewTarget(t oras.Target, actionPrompt, donePrompt string, tty *os.File) (*Target, error) {
+	manager, err := progress.NewManager(tty)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Target{
+		Target:       t,
+		manager:      manager,
+		actionPrompt: actionPrompt,
+		donePrompt:   donePrompt,
+	}, nil
+}
+
+// Push pushes the content to the Target with tracking.
+func (t *Target) Push(ctx context.Context, expected ocispec.Descriptor, content io.Reader) error {
+	r, err := managedReader(content, expected, t.manager, t.actionPrompt, t.donePrompt)
+	if err != nil {
+		return err
+	}
+	defer r.Close()
+	r.Start()
+	if err := t.Target.Push(ctx, expected, r); err != nil {
+		return err
+	}
+	r.Done()
+	return nil
+}
+
+// PushReference pushes the content to the Target with tracking.
+func (t *Target) PushReference(ctx context.Context, expected ocispec.Descriptor, content io.Reader, reference string) error {
+	r, err := managedReader(content, expected, t.manager, t.actionPrompt, t.donePrompt)
+	if err != nil {
+		return err
+	}
+	defer r.Close()
+	r.Start()
+	if rp, ok := t.Target.(registry.ReferencePusher); ok {
+		err = rp.PushReference(ctx, expected, r, reference)
+	} else {
+		if err := t.Target.Push(ctx, expected, r); err != nil {
+			return err
+		}
+		err = t.Target.Tag(ctx, expected, reference)
+	}
```

Comment:
If the embeded `Target` is not a `registry.ReferencePusher`, we should implement it. In other words, we should not have the `func (t *Target) PushReference` method in the first place.

Reply:
We can discuss the coding trick offline.